lib/puppet/application/resource.rb

```diff
@@ -155,7 +155,9 @@ def main
 
       if options[:to_yaml]
         data = resources.map do |resource|
-          resource.prune_parameters(:parameters_to_include => @extra_params).to_hiera_hash
+          Puppet.override(rich_data: false) do
+            resource.prune_parameters(:parameters_to_include => @extra_params).to_hiera_hash
+          end
```

Comment:
`to_hiera_hash` calls `to_data_hash` which I would expect to serialize rich data correctly (since the base context pushes `rich_data` onto the context)

https://github.com/puppetlabs/puppet/blob/7cc46f776d836fe5d07148f3ab6865f4f9e05bbc/lib/puppet/resource.rb#L469

What was the error you ran into?

Reply:
Oh I see, since the `string` property returns a `ProcessOutput` which subclasses `String` rich data serialization produces 

```
       -    string: test
       +    string:
       +      __ptype: String
       +      __pvalue: test
```

So we want to preserve "non-rich" data for the `puppet resource <type> ...` command

Reply:
I'm not sure if we should be disabling rich_data like

```ruby
Puppet.override(rich_data: false) do
  resource.prune_parameters(:parameters_to_include => @extra_params).to_hiera_hash
end
```

Or if we should enable `stringify_rich` like we do in https://github.com/puppetlabs/puppet/blob/7bd9221527e3f6481ae353359e951a88fb5c6eee/lib/puppet/application/resource.rb#L235

Ultimately I think we want to get to a point where rich data is always enabled throughout puppet so that we deprecate "non-rich data" serialization. In situations where we need a lossy-stringified representation, like `puppet resource --to_yaml ...`, then we can use `stringify_rich = true`. Thoughts?

Also wondering if we should lift `stringify_rich => false` to the outer override so it's enabled throughout the `resource` application

```ruby
Puppet.override(:current_environment => env, :loaders => Puppet::Pops::Loaders.new(env), stringify_rich: false) do
```

Reply:
> ...then we can use `stringify_rich = true`. Thoughts?

I think that sounds good. I was unclear exactly what `stringify_rich` implied when I first saw it. What would you think about taking the default block from the Puppet.lookup call and added the default `stringify_rich: false` to the base_context?

Also, setting `stringify_rich: true` in this override causes this test case: https://github.com/puppetlabs/puppet/pull/9471/files#diff-4775757bedef0d6fce0f8e35ccc133c1937c80bb68db53b1899488f6b68c8667R173 to fail whether rich_data is false or not. How important is the object failing to be serialized (is it harmful to lossily serialize arbitrary objects like that?

Reply:
> Also wondering if we should lift `stringify_rich => false` to the outer override so it's enabled throughout the resource application

Do you mean `stringify_right => true`?

Reply:
I've updated this PR with what I think you're getting at, plus my query about moving stringify_rich to base_context. let me know if I understood it correctly and if the update to base_context is okay or not.

Reply:
> How important is the object failing to be serialized (is it harmful to lossily serialize arbitrary objects like that?

That test came about because the mysql module was serializing `P::U::E::ProcessOutput` objects (which extend from `String`). So the error was a sign that the module's provider needed to call `ProcessOutput#to_s`.

But that's at odds with using `stringify_rich` to lossily convert the output. I think that's ok...

> Do you mean stringify_right => true?

Yeah sorry, in the `puppet resource` case we want `stringify_rich => true`

> if the update to base_context is okay or not.

Yeah I think setting the default for `stringify_rich` in the base context is better than the way it "magically appears" in `lib/puppet/resource.rb`.

Is it guaranteed that applications using puppet as a library, like bolt, rspec-puppet, module CI, will initialize the base_context (so that `stringify_rich` is defined)? If not, then removing the default from `lib/puppet/resource.rb` will cause an exception since no value would be defined.

Reply:
> Is it guaranteed that applications using puppet as a library, like bolt, rspec-puppet, module CI, will initialize the base_context

That's my understanding. It's initialized conditionally as part of [`Puppet.initialize_settings`](https://github.com/puppetlabs/puppet/blob/main/lib/puppet.rb#L177) (the conditional part was added so that Puppet Server could not do it for every thread in multithreaded mode).

Double checking I see that rspec-puppet calls into the [`Puppet::Test::TestHelper.initialize`](https://github.com/rodjek/rspec-puppet/blob/c4ceecde20c9c53a388587dd37d6a2fc5dac50e5/lib/rspec-puppet.rb#L82) which sets up the base_context.

I see Bolt explicitly calling Puppet.initialize_settings in three places, though I'm unsure if it calls it everywhere it might serialize a resource. I'll ask around more.

Reply:
I've asked the maintainers of Bolt to confirm, but I believe [this](https://github.com/puppetlabs/bolt/blob/main/lib/bolt/pal.rb#L119) means that it never instantiates PAL without initializing the Puppet settings.

Reply:
Yes, we have been very careful to ensure settings initialization happens whenever we use pal.